api/v1alpha1/addresspool_webhook.go

```diff
@@ -0,0 +1,168 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/mikioh/ipaddr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging addresspool-webhook.
+var addresspoollog = logf.Log.WithName("addresspool-webhook")
+var c client.Client
+
+func (addressPool *AddressPool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	c = mgr.GetClient()
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(addressPool).
+		Complete()
+}
+
+//+kubebuilder:webhook:verbs=create;update;delete,path=/validate-metallb-io-v1alpha1-addresspool,mutating=false,failurePolicy=fail,groups=metallb.io,resources=addresspools,versions=v1alpha1,name=addresspoolvalidationwebhook.metallb.io
+var _ webhook.Validator = &AddressPool{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateCreate() error {
+	addresspoollog.Info("validate AddressPool creation", "name", addressPool.Name)
+
+	return addressPool.validateAddressPool(true)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateUpdate(old runtime.Object) error {
+	addresspoollog.Info("validate AddressPool update", "name", addressPool.Name)
+
+	return addressPool.validateAddressPool(false)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for AddressPool.
+func (addressPool *AddressPool) ValidateDelete() error {
+	addresspoollog.Info("validate AddressPool deletion", "name", addressPool.Name)
+
+	return nil
+}
+
+func cidrsOverlap(a, b *net.IPNet) bool {
+	return cidrContainsCIDR(a, b) || cidrContainsCIDR(b, a)
+}
+
+func cidrContainsCIDR(outer, inner *net.IPNet) bool {
```

Comment:
Isn't enough to say `outer.Contains(inner.IP)` ?

Reply:
this code is taken from metallb config, if we ever move the crds and the webhook to metallb we would like to use the same code. so for now let's keep it as it is.